File: project-analyzer-ts/src/plugins/generic-plugin.ts
/**
 * GenericPlugin — 通用语言插件（基于文件模式匹配）
 *
 * 当没有对应语言插件时使用此插件，提供基础的文件分析能力：
 * - parseFile: 使用简单正则提取函数/类名
 * - extractDependencies: 返回空数组
 * - identifyApis: 返回空数组
 * - identifyModules: 基于顶层目录结构识别模块
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import type { LanguagePlugin } from './language-plugin.js';
import type { AstNode, Dependency, ApiEndpoint, ModuleInfo } from '../models/index.js';

export class GenericPlugin implements LanguagePlugin {
  getLanguageId(): string {
    return 'generic';
  }

  parseFile(filePath: string): AstNode[] {
    let content: string;
    try {
      content = fs.readFileSync(filePath, 'utf-8');
    } catch {
      return [];
    }

    const lines = content.split('\n');
    const totalLines = lines.length;
    const nodes: AstNode[] = [];

    // Extract class-like declarations
    const classRegex = /^\s*(?:export\s+)?(?:abstract\s+)?(?:class|struct)\s+(\w+)/;
    // Extract function/method declarations (various styles)
    const funcRegex = /^\s*(?:export\s+)?(?:async\s+)?(?:function\s+|def\s+|func\s+|fn\s+)(\w+)/;

    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];

      const classMatch = line.match(classRegex);
      if (classMatch) {
        nodes.push({
          type: 'class',
          name: classMatch[1],
          filePath,
          startLine: i + 1,
          endLine: i + 1,
          modifiers: [],
          annotations: [],
          children: [],
        });
        continue;
      }

      const funcMatch = line.match(funcRegex);
      if (funcMatch) {
        nodes.push({
          type: 'function',
          name: funcMatch[1],
          filePath,
          startLine: i + 1,
          endLine: i + 1,
          modifiers: [],
          annotations: [],
          children: [],
        });
      }
    }

    // If no specific constructs found, return a file-level module node
    if (nodes.length === 0) {
      nodes.push({
        type: 'module',
        name: path.basename(filePath),
        filePath,
        startLine: 1,
        endLine: totalLines,
        modifiers: [],
        annotations: [],
        children: [],
      });
    }

    return nodes;
  }

  extractDependencies(_projectRoot: string): Dependency[] {
    return [];
  }

  identifyApis(_filePath: string): ApiEndpoint[] {
    return [];
  }

  identifyModules(projectRoot: string): ModuleInfo[] {
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(projectRoot, { withFileTypes: true });
    } catch {
      return [];
    }

    const modules: ModuleInfo[] = [];

    for (const entry of entries) {
      if (!entry.isDirectory()) continue;

      const dirName = entry.name;
      // Skip hidden dirs and common non-module dirs
      if (dirName.startsWith('.') || IGNORED_DIRS.has(dirName)) continue;

      const dirPath = path.join(projectRoot, dirName);
      const keyFiles = listSourceFiles(dirPath);

      modules.push({
        name: dirName,
        path: dirPath,
        description: `基于目录推断的模块: ${dirName}`,
        isInferred: true,
        keyClasses: [],
        keyFiles,
        dependencies: [],
      });
    }

    return modules;
  }
}

/** Directories to skip when identifying modules */
const IGNORED_DIRS = new Set([
  'node_modules', 'dist', 'build', 'out', 'target',
  '.git', '.svn', '.hg',
  '__pycache__', '.gradle', '.idea', '.vscode',
  'vendor', 'coverage', '.next', '.nuxt',
]);

/**
 * List source files in a directory (non-recursive, top-level only).
 */
function listSourceFiles(dirPath: string): string[] {
  try {
    const entries = fs.readdirSync(dirPath, { withFileTypes: true });
    return entries
      .filter((e) => e.isFile() && !e.name.startsWith('.'))
      .map((e) => path.join(dirPath, e.name));
  } catch {
    return [];
  }
}

File: project-analyzer-java/src/main/java/com/analyzer/plugin/GenericPlugin.java
package com.analyzer.plugin;

import com.analyzer.model.*;

import java.io.IOException;
import java.nio.file.*;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.*;
import java.util.regex.*;
import java.util.stream.Stream;

public class GenericPlugin implements LanguagePlugin {

    private static final Set<String> IGNORED_DIRS = Set.of(
            "node_modules", ".git", "dist", "build", "__pycache__",
            "target", "vendor", ".gradle", ".idea", ".vscode",
            ".next", ".nuxt", "coverage", ".tox", "venv", ".venv"
    );

    private static final Pattern CLASS_PATTERN = Pattern.compile(
            "(?:public\\s+|private\\s+|protected\\s+)?(?:abstract\\s+)?class\\s+(\\w+)");
    private static final Pattern FUNCTION_PATTERN = Pattern.compile(
            "(?:public|private|protected|static|final|\\s)+[\\w<>\\[\\]]+\\s+(\\w+)\\s*\\(");

    @Override
    public String getLanguageId() { return "generic"; }

    @Override
    public List<AstNode> parseFile(Path filePath) {
        List<AstNode> nodes = new ArrayList<>();
        try {
            List<String> lines = Files.readAllLines(filePath);
            String fileStr = filePath.toString();
            for (int i = 0; i < lines.size(); i++) {
                String line = lines.get(i);
                Matcher cm = CLASS_PATTERN.matcher(line);
                if (cm.find()) {
                    AstNode node = new AstNode();
                    node.setType(AstNode.AstNodeType.CLASS);
                    node.setName(cm.group(1));
                    node.setFilePath(fileStr);
                    node.setStartLine(i + 1);
                    node.setEndLine(i + 1);
                    node.setModifiers(new ArrayList<>());
                    node.setAnnotations(new ArrayList<>());
                    node.setChildren(new ArrayList<>());
                    nodes.add(node);
                }
                Matcher fm = FUNCTION_PATTERN.matcher(line);
                if (fm.find() && !line.contains("class ")) {
                    AstNode node = new AstNode();
                    node.setType(AstNode.AstNodeType.METHOD);
                    node.setName(fm.group(1));
                    node.setFilePath(fileStr);
                    node.setStartLine(i + 1);
                    node.setEndLine(i + 1);
                    node.setModifiers(new ArrayList<>());
                    node.setAnnotations(new ArrayList<>());
                    node.setChildren(new ArrayList<>());
                    nodes.add(node);
                }
            }
        } catch (IOException ignored) {}
        return nodes;
    }

    @Override
    public List<Dependency> extractDependencies(Path projectRoot) {
        return Collections.emptyList();
    }

    @Override
    public List<ApiEndpoint> identifyApis(Path filePath) {
        return Collections.emptyList();
    }

    @Override
    public List<ModuleInfo> identifyModules(Path projectRoot) {
        List<ModuleInfo> modules = new ArrayList<>();
        try (DirectoryStream<Path> stream = Files.newDirectoryStream(projectRoot)) {
            for (Path entry : stream) {
                if (Files.isDirectory(entry)) {
                    String name = entry.getFileName().toString();
                    if (IGNORED_DIRS.contains(name) || name.startsWith(".")) continue;
                    List<String> keyFiles = collectSourceFiles(entry);
                    if (!keyFiles.isEmpty()) {
                        modules.add(new ModuleInfo(name, entry.toString(),
                                "基于目录推断的模块: " + name, true,
                                new ArrayList<>(), keyFiles, new ArrayList<>()));
                    }
                }
            }
        } catch (IOException ignored) {}
        return modules;
    }

    private List<String> collectSourceFiles(Path dir) {
        List<String> files = new ArrayList<>();
        try (Stream<Path> walk = Files.walk(dir)) {
            walk.filter(Files::isRegularFile)
                .filter(p -> {
                    String name = p.getFileName().toString();
                    return name.endsWith(".java") || name.endsWith(".py") || name.endsWith(".ts")
                            || name.endsWith(".js") || name.endsWith(".go") || name.endsWith(".rb");
                })
                .forEach(p -> files.add(p.toString()));
        } catch (IOException ignored) {}
        return files;
    }
}
